### PR TITLE
CSHARP-4778: Add synonym support to text operator for Atlas Search

### PR DESCRIPTION
**Pull request by adelinowona**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs [hunk 1]
```diff
@@ -477,6 +477,133 @@ public void Text()
...
+        [Fact]
+        public void TextQueryWithEmptySynonymsMapNameShouldThrowException()
+        {
+            var exception = Record.Exception(() => Builders<Movie>.Search.Text("title", "vehicle", ""));
+
+            exception.Should().BeOfType<ArgumentException>();
+            exception.Message.Should().StartWith("Value cannot be empty.");
+        }
+
+
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:524]: minor: remove extra blank line.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs [hunk 2]
```diff
@@ -477,6 +477,133 @@ public void Text()
...
+            var sailSearchResults =
+                SearchSingleSynonymMapping(Builders<Movie>.Search.Text(x => x.Title, "sail",
+                    "transportSynonyms"));
+
+            boatSearchResults.Should().NotBeNull();
+            vesselSearchResults.Should().NotBeNull();
+            sailSearchResults.Should().NotBeNull();
+
+            Assert.False(boatSearchResults.SequenceEqual(vesselSearchResults));
+            Assert.False(vesselSearchResults.SequenceEqual(sailSearchResults));
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:585]: `AtlasSearchTests` are more like a smoke tests, and they don't have to be exhaustive. It's usually single test method per parameter. Let's try to think how to compress those test cases into single method or two.

File: tests/MongoDB.Driver.Tests/Search/SearchDefinitionBuilderTests.cs [hunk 3]
```diff
@@ -195,17 +195,17 @@ public void EmbeddedDocument()
             // Compound
             var compound = subject
                 .Compound()
-                .Must(subject.Text("y", "1"))
+                .Must(subject.Text("y", "1", "testSynonyms"))
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Search/SearchDefinitionBuilderTests.cs:198]: Move synonym tests to `Text` and `Text_typed` (instead of `Emebedded`)

**Review by BorisDog — CHANGES_REQUESTED** [review]
Looks great overall! Some changes requested.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs [hunk 4]
```diff
@@ -477,6 +477,52 @@ public void Text()
...
+            var vesselAndFedoraSearchResults = SearchMultipleSynonymMapping(
+                Builders<Movie>.Search.Text(x => x.Title, "vessel", "transportSynonyms"),
+                Builders<Movie>.Search.Text(x => x.Title, "fedora", "attireSynonyms"));
+
+            automobileAndAtttireSearchResults.Should().NotBeNull();
+            vehicleAndDressSearchResults.Should().NotBeNull();
+            boatAndHatSearchResults.Should().NotBeNull();
+            vesselAndFedoraSearchResults.Should().NotBeNull();
+
+            Assert.True(automobileAndAtttireSearchResults.SequenceEqual(vehicleAndDressSearchResults));
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:522]: Should we use fluent assertions here?

**Review by sanych-sun — APPROVED** [review]
LGTM + minor comment

File: src/MongoDB.Driver/Search/OperatorSearchDefinitions.cs [hunk 5]
```diff
@@ -399,23 +399,27 @@ internal sealed class TextSearchDefinition<TDocument> : OperatorSearchDefinition
...
             _fuzzy = fuzzy;
+            _synonyms = synonyms;
         }
 
         private protected override BsonDocument RenderArguments(SearchDefinitionRenderContext<TDocument> renderContext) =>
             new()
             {
                 { "query", _query.Render() },
                 { "fuzzy", () => _fuzzy.Render(), _fuzzy != null },
+                { "synonyms", () => new BsonString(_synonyms), _synonyms != null }
```
Comment (BorisDog) [inline src/MongoDB.Driver/Search/OperatorSearchDefinitions.cs:422]: `{ "synonyms", _synonyms, _synonyms != null }` should also work.

File: src/MongoDB.Driver/Search/SearchDefinitionBuilder.cs [hunk 6]
```diff
@@ -627,7 +627,23 @@ public sealed class SearchDefinitionBuilder<TDocument>
...
+        /// <param name="query">The string or strings to search for.</param>
+        /// <param name="synonyms">The name of the synonym mapping definition in the index definition</param>
+        /// <param name="score">The score modifier.</param>
+        /// <returns>A text search definition.</returns>
+        public SearchDefinition<TDocument> Text(
+            SearchPathDefinition<TDocument> path,
+            SearchQueryDefinition query,
+            string synonyms,
+            SearchScoreDefinition<TDocument> score = null) =>
+            new TextSearchDefinition<TDocument>(path, query, null, score, synonyms);
```
Comment (BorisDog) [inline src/MongoDB.Driver/Search/SearchDefinitionBuilder.cs:646]: Match indentation of other overloads.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs [hunk 7]
```diff
@@ -477,6 +477,52 @@ public void Text()
...
+        {
+            var sortDefinition = Builders<Movie>.Sort.Ascending(x => x.Title);
+            var result =
+                GetSynonymTestCollection().Aggregate()
+                    .Search(Builders<Movie>.Search.Text(x => x.Title, query, synonym), indexName: "synonyms-tests")
+                    .Sort(sortDefinition)
+                    .Project<Movie>(Builders<Movie>.Projection.Include("Title").Exclude("_id"))
+                    .Limit(1)
+                    .ToList()
+                    .Single();
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:493]: We should use Single directly. `Limit(1).Single()`
Also in the old code as well.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs [hunk 8]
```diff
@@ -477,6 +477,52 @@ public void Text()
...
+
+            result.Title.Should().Be(expected);
+        }
+
+        [Fact]
+        public void TextWithSynonymsMappings()
+        {
+            var automobileAndAtttireSearchResults = SearchMultipleSynonymMapping(
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:501]: `automobileAndAttireSearchResults`

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs (hunk 4 above)
Comment (adelinowona) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTests.cs:522]: Agreed, helps keep consistency in tests. Thanks!

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**